pkg/virtctl/adm/logverbosity/logverbosity.go

```diff
@@ -0,0 +1,451 @@
+	usage += "  {{ProgramName}} adm log-verbosity --virt-handler=7\n"
+
+	usage += "  # show log-verbosity for virt-handler and virt-launcher\n"
+	usage += "  {{ProgramName}} adm log-verbosity --virt-handler --virt-launcher\n"
+	usage += "  # set log-verbosity for virt-handler to 7 and virt-launcher to 3\n"
+	usage += "  {{ProgramName}} adm log-verbosity --virt-handler=7 --virt-launcher=3\n"
+
+	usage += "  # reset all components to default besides virt-handler which is 7\n"
+	usage += "  {{ProgramName}} adm log-verbosity --reset --virt-handler=7\n"
+	usage += "  # set all components to 3 besides virt-handler which is 7\n"
+	usage += "  {{ProgramName}} adm log-verbosity --all=3 --virt-handler=7\n"
+
+	return usage
+}
+
+// VirtComponent to component name
+// also used by the test file
+func GetComponentNameByVirtComponent(component virtComponent) string {
+	var virtComponentToComponentName = map[virtComponent]string{
+		VirtAPI:        "virt-api",
+		VirtController: "virt-controller",
+		VirtHandler:    "virt-handler",
+		VirtLauncher:   "virt-launcher",
+		VirtOperator:   "virt-operator",
+		All:            "all",
+	}
+	return virtComponentToComponentName[component]
+}
+
+// virtComponent to JSON name
+func getJSONNameByVirtComponent(component virtComponent) string {
+	var virtComponentToJSONName = map[virtComponent]string{
+		VirtAPI:        "virtAPI",
+		VirtController: "virtController",
+		VirtHandler:    "virtHandler",
+		VirtLauncher:   "virtLauncher",
+		VirtOperator:   "virtOperator",
+		All:            "all",
+	}
+	return virtComponentToJSONName[component]
+}
+
+// component name to JSON name
+func getJSONNameByComponentName(componentName string) string {
+	var componentNameToJSONName = map[string]string{
+		"virt-api":        "virtAPI",
+		"virt-controller": "virtController",
+		"virt-handler":    "virtHandler",
+		"virt-launcher":   "virtLauncher",
+		"virt-operator":   "virtOperator",
+		"all":             "all",
+	}
+	return componentNameToJSONName[componentName]
+}
+
+func detectInstallNamespaceAndName(virtClient kubecli.KubevirtClient) (namespace, name string, err error) {
+	// listing KubeVirt CRs in all namespaces and see where it is installed
+	kvs, err := virtClient.KubeVirt("").List(&k8smetav1.ListOptions{})
+	if err != nil {
+		return "", "", fmt.Errorf("could not list KubeVirt CRs in all namespaces: %v", err)
+	}
+	if len(kvs.Items) == 0 {
+		return "", "", errors.New("could not detect a KubeVirt installation")
+	}
+	if len(kvs.Items) > 1 {
+		return "", "", errors.New("invalid kubevirt installation, more than one KubeVirt resource found")
+	}
+	namespace = kvs.Items[0].Namespace
+	name = kvs.Items[0].Name
+	return
+}
+
+func hasVerbosityInKV(kv *v1.KubeVirt) (map[string]uint, error) {
+	verbosityMap := map[string]uint{} // key: component name, value: verbosity
+	// check the logVerbosity field in the KubeVirt CR
+	lvJSON, err := json.Marshal(kv.Spec.Configuration.DeveloperConfiguration.LogVerbosity)
```

Comment:
Can any of the field here be a pointer and be equal `nil`?

Reply:
Great point! This is very tricky. Someone found it too.
https://github.com/golang/go/issues/48646

The manual mentions:
- The JSON null value unmarshals into an interface, map, pointer, or slice by setting that Go value to nil. Because null is often used in JSON to mean “not present,” unmarshaling a JSON null into any other Go type has no effect on the value and produces no error.

https://pkg.go.dev/encoding/json#Unmarshal

Here is an illustration using our code.
```
If kv.Spec.Configuration.DeveloperConfiguration.LogVerbosity = nil (LogVerbosity absent)
Then, lvJSON = null (not {})
Therefore, verbosityMap = nil (not map[])
```
Even though we've initialized `verbosityMap`, `Unmarshal` returns an uninitialized map (nil).

I guess people might expect `Unmarshal` to do nothing with `verbosityMap`, i.e. just return `verbosityMap` as is (which is already initialized). But this is not the case.

On the other hand, 
```
If kv.Spec.Configuration.DeveloperConfiguration.LogVerbosity = {} (LogVerbosity empty)
Then, lvJSON = {}
Therefore, verbosityMap = map[]
```
In this case, `Unmarshal` returns `verbosityMap` as is (initialized map).

Reply:
I was referring even before the marshalling if for example `DeveloperConfiguration` can be nil

Reply:
I think it depends on how the KubeVirt CR is initialized. **It seems you are right**, which means `kv.Spec.Configuration.DeveloperConfiguration.LogVerbosity` might never be `nil`. But, **I am not 100% sure**. Here is my investigation.

1. The following statement is taken from the document below. From this statement, it appears that manually setting `LogVerbosity` is optional, and then `LogVerbosity` can be `nil`.

- One way of raising log verbosity is to manually determine it for the different components in KubeVirt CR:
https://kubevirt.io/user-guide/operations/debug/#setting-verbosity-per-kubevirt-component

2. According to the definition of the `DeveloperConfiguration` struct below, LogVerbosity can be empty (i.e. optional). So, it seems `kv.Spec.Configuration.DeveloperConfiguration.LogVerbosity` can be `nil`.
https://github.com/kubevirt/kubevirt/blob/main/staging/src/kubevirt.io/api/core/v1/types.go
```
type DeveloperConfiguration struct {
	…
	LogVerbosity               *LogVerbosity     `json:"logVerbosity,omitempty"`
}
```

3. I also checked the yaml file (`kubevirt-cr.yaml`) provided for the KubeVirt installation. It seems the `LogVerbosity` is absent initially.
https://github.com/kubevirt/kubevirt/releases/download/v1.0.0/kubevirt-cr.yaml
```
apiVersion: kubevirt.io/v1
kind: KubeVirt
metadata:
  name: kubevirt
  namespace: kubevirt
spec:
  certificateRotateStrategy: {}
  configuration:
    developerConfiguration:
      featureGates: []
  customizeComponents: {}
  imagePullPolicy: IfNotPresent
  workloadUpdateStrategy: {}
```

So far, it seems `kv.Spec.Configuration.DeveloperConfiguration.LogVerbosity` could be `nil`.

4. However, looking at the following code, which is supposed to be used by each KubeVirt component to get the CR (configuration filed in the CR), `defaultClusterConfig()` gets the default config, then `setConfigFromKubeVirt()` writes the contents of the actual CR on top of the default config.
https://github.com/kubevirt/kubevirt/blob/main/pkg/virt-config/configuration.go#L337
Here, items that exist in the default config but do not exist in actual CR remain as they are in default. Since LogVerbosity is included in the default config obtained by `defaultClusterConfig()` (below), `LogVerbosity` exists even if `LogVerbosity` is not included in the actual CR.
https://github.com/kubevirt/kubevirt/blob/main/pkg/virt-config/configuration.go#L158

If this (4) understanding is correct, then `kv.Spec.Configuration.DeveloperConfiguration.LogVerbosity` should never be `nil` (always present). I do not know the entire codebase, so **I cannot say that the above understanding is always correct**.

If this understanding is correct, then the following code should be deleted (in addition, the test code for the case where `logVerbosity` is `nil` should also be deleted).
```
if verbosityMap == nil {
	verbosityMap = map[string]uint{}
}
```

Since if `kv.Spec.Configuration.DeveloperConfiguration.LogVerbosity` is `nil`, `panic: assignment to entry in nil map` will occur, and the presence of this code does not cause any side effects, I think **it would be safer to leave the above `if clause` if we are not 100% sure that the above understanding is correct (i.e. there is zero chance of **nil**)**.

Do you know if the above understanding is correct? In other words, are you sure that `kv.Spec.Configuration.DeveloperConfiguration.LogVerbosity` will never be `nil`? If so, I’ll remove the above `if closure`, and also change the test code.

Please let me know your opinion.

Reply:
I may have misunderstood 4 in the previous comment. This is the way each KubeVirt component handles the KubeVirt CR. It does not necessarily mean that the KubeVirt CR always includes the `LogVerbosity`. So, I think the `LogVerbosity` in the KubeVirt CR could be nil (cannot say the `LogVerbosity` is always present).

I checked more like following.

1. I ran `make cluster-up` and `make cluster-sync`, and checked the KubeVirt CR. I found the `LogVerbosity` is absent in the KubeVirt CR.
`~/kubevirt/cluster-up/kubectl.sh describe kubevirt -n kubevirt`

2. I had a chance to see the installed OpenShift virtualization environment. I do not know much about OpenShift virtualization, but I guess the KubeVirt CR is installed in the `openshift-cnv` namespace and named `kubevirt-kubevirt-hyperconverged`. I looked at it and it had the `DeveloperConfiguration`, but not the `LogVerbosity`.

**In summary, I think it would be safer to assume that `LogVerbosity` could be nil.**

I changed the code as follows. I think this change makes the logic straightforward and easy to maintain in the future.
```
func hasVerbosityInKV(kv *v1.KubeVirt) (map[string]uint, error) {
	verbosityMap := map[string]uint{} // key: component name, value: verbosity

	// if LogVerbosity is not present in the KubeVirt CR, skip here
	if kv.Spec.Configuration.DeveloperConfiguration.LogVerbosity != nil {
		// get the logVerbosity field in the KubeVirt CR, and put it to verbosityMap
		lvJSON, err := json.Marshal(kv.Spec.Configuration.DeveloperConfiguration.LogVerbosity)
		if err != nil {
			return nil, err
		}
		if err := json.Unmarshal(lvJSON, &verbosityMap); err != nil {
			return nil, err
		}
	}

	return verbosityMap, nil
}
```

I hope this makes sense to you.